src/main.rs

```diff
@@ -1,42 +1,71 @@
-use nix::sys::sysinfo::{sysinfo, SysInfo};
-
-extern crate num_cpus;
 
 fn main() -> io::Result<()> {
-    let cpu_count = num_cpus::get();
-    println!("cpu count: {}", cpu_count);
-    let mut buffer = [0; 1024];
     let mut stdout = StandardStream::stdout(ColorChoice::Always);
+    let mut buffer = String::new();
+    let mut stdin = BufReader::new(stdin());
+    let mut last_load = 0.0;
 
-    // Set the color of the output to red
-    stdout.set_color(ColorSpec::new().set_fg(Some(Color::Red)))?;
+    // with sampling interval of 100 ms this amounts to 1000 seconds of data
+    // if it takes so long to produce a line of data, then the cpu load does not really matter
+    let (cpu_tx, cpu_rx) = sync_channel(100_000);
 
-    // Read input into the buffer in a loop until we reach the end of the input stream
-    loop {
-        let bytes_read = io::stdin().read(&mut buffer)?;
+    // This thread will provide a stream of cpu measurement values
+    std::thread::spawn(move || {
+        let mut sys = System::new();
+        sys.refresh_cpu();
 
-        // If we didn't read any bytes, we've reached the end of the input
-        if bytes_read == 0 {
-            break;
+        loop {
+            if let Err(TrySendError::Disconnected(..)) =
+                cpu_tx.try_send(sys.global_cpu_info().cpu_usage())
+            {
+                break;
+            }
+            std::thread::sleep(Duration::from_millis(100));
+            sys.refresh_cpu();
         }
+    });
 
-        // Get the current system information
-        let sys_info: SysInfo = sysinfo().unwrap();
-        // Calculate the current (1 minute average) CPU load as a percentage
-        let cpu_load = sys_info.load_average().0 as f32 / cpu_count as f32 * 100.0;
-        println!("cpu load: {}", cpu_load);
-        // Calculate the color for the current CPU load using a gradient function
-        let color = get_color(cpu_load / 100.0);
-        // Set the color of the output based on the CPU load
-        stdout.set_color(ColorSpec::new().set_fg(Some(color)))?;
-        // Write the bytes we read to standard output
-        io::stdout().write(&buffer[..bytes_read])?;
+    loop {
+        match stdin.read_line(&mut buffer) {
+            Ok(0) => exit(0),
+            Ok(_) => {
+                print_colored(&cpu_rx, &buffer, &mut stdout, &mut last_load);
+                buffer.clear();
+            }
+            Err(e) => {
+                println!("{e}");
+                exit(1)
+            }
+        }
     }
+}
 
-    Ok(())
+fn print_colored(
+    cpu_rx: &std::sync::mpsc::Receiver<f32>,
+    buffer: &str,
+    stdout: &mut StandardStream,
+    last_load: &mut f32,
+) {
+    let cpu_loads = {
+        let mut result = vec![];
+        while let Ok(v) = cpu_rx.try_recv() {
+            result.push(v)
+        }
+        result
+    };
+    if !cpu_loads.is_empty() {
+        *last_load = cpu_loads.iter().copied().sum::<f32>() / cpu_loads.len() as f32;
```

Comment:
for me lines that have more that 2 expressions are complex enough to have a comment

Reply:
Added